File: mobile/android/app/src/main/java/com/volttracker/obdpoc/ObdElmDecode.java
package com.volttracker.obdpoc;

import com.volttracker.obdpoc.data.ObdLocalStore;
import java.util.Locale;

/**
 * Stateless decoding, classification and formatting helpers for the OBD/ELM layer. Extracted from
 * {@link ObdService} so the service file stays focused and these pure functions are unit-testable
 * on their own. {@code ObdService} static-imports this class.
 */
final class ObdElmDecode {

    private ObdElmDecode() {}

    static boolean hasElmPrompt(String response) {
        return response != null && response.indexOf('>') >= 0;
    }

    static String appendRaw(String raw, String command, String response) {
        return raw + command + ": " + summarizeForStorage(command, response) + "\n";
    }

    static String summarizeForStorage(String command, String response) {
        String summary = ObdProtocol.summarize(response);
        if (!isVinCommand(command)) {
            return summary;
        }
        if (summary.isEmpty()) {
            return "";
        }
        return "[VIN redacted; responseLength=" + summary.length() + "]";
    }

    private static boolean isVinCommand(String command) {
        return command != null && "0902".equals(command.trim().toUpperCase(Locale.US));
    }

    static String pidForCommand(String command) {
        if (command == null) {
            return "";
        }
        String clean = command.trim().toUpperCase(Locale.US);
        if (clean.startsWith("01") && clean.length() >= 4) {
            return clean.substring(2, 4);
        }
        if (clean.startsWith("22") && clean.length() >= 6) {
            return clean.substring(2);
        }
        if (clean.startsWith("09") && clean.length() >= 4) {
            return clean.substring(2, 4);
        }
        return "";
    }

    static String nameForCommand(String command) {
        if (command == null) {
            return "";
        }
        String clean = command.trim().toUpperCase(Locale.US);
        if ("ATRV".equals(clean)) {
            return "adapter voltage";
        }
        if ("010D".equals(clean)) {
            return "vehicle speed";
        }
        if ("010C".equals(clean)) {
            return "engine rpm";
        }
        if ("0105".equals(clean)) {
            return "coolant temperature";
        }
        if ("010F".equals(clean)) {
            return "intake air temperature";
        }
        if ("0104".equals(clean)) {
            return "engine load";
        }
        if ("0111".equals(clean)) {
            return "throttle position";
        }
        if ("0149".equals(clean)) {
            return "accelerator pedal position";
        }
        if ("0142".equals(clean)) {
            return "control module voltage";
        }
        if ("011F".equals(clean)) {
            return "engine run time";
        }
        if ("01A6".equals(clean)) {
            return "odometer";
        }
        if ("012F".equals(clean)) {
            return "fuel level";
        }
        if ("015C".equals(clean)) {
            return "engine oil temperature";
        }
        if ("22119F".equals(clean) || "22119F01".equals(clean)) {
            return "engine oil life";
        }
        if ("221154".equals(clean)) {
            return "engine oil temperature";
        }
        if ("22203F".equals(clean)) {
            return "engine torque";
        }
        if ("015B".equals(clean)) {
            return "state of charge";
        }
        if ("222429".equals(clean)) {
            return "hv pack voltage";
        }
        if ("222414".equals(clean)) {
            return "hv pack current";
        }
        if ("222883".equals(clean)) {
            return "motor A current";
        }
        if ("222884".equals(clean)) {
            return "motor B current";
        }
        if ("222885".equals(clean)) {
            return "motor A voltage";
        }
        if ("222886".equals(clean)) {
            return "motor B voltage";
        }
        if ("222487".equals(clean)) {
            return "ev distance this cycle";
        }
        if ("222889".equals(clean)) {
            return "prndl state";
        }
        if ("221940".equals(clean) || "22194001".equals(clean)) {
            return "transmission temperature";
        }
        if ("22434F".equals(clean)) {
            return "hv battery temperature";
        }
        if ("224368".equals(clean)) {
            return "charger ac voltage";
        }
        if ("224369".equals(clean)) {
            return "charger ac current";
        }
        if ("22436B".equals(clean)) {
            return "charger hv voltage";
        }
        if ("22436C".equals(clean)) {
            return "charger hv current";
        }
        if ("224373".equals(clean)) {
            return "charging mode";
        }
        if ("22437D".equals(clean)) {
            return "last charge energy";
        }
        if ("2243A5".equals(clean)) {
            return "hv battery charge count";
        }
        if ("2243AF".equals(clean)) {
            return "hv battery raw soc";
        }
        if ("224531".equals(clean)) {
            return "charging level";
        }
        if ("228334".equals(clean)) {
            return "hv battery displayed soc";
        }
        if ("2241B2".equals(clean)) {
            return "battery coolant pump rpm";
        }
        if ("2241B4".equals(clean)) {
            return "battery coolant valve";
        }
        if ("2241B6".equals(clean)) {
            return "battery heater power";
        }
        if ("22801E".equals(clean)) {
            return "outside air temperature raw";
        }
        if ("22801F".equals(clean)) {
            return "outside air temperature filtered";
        }
        if ("22248E".equals(clean)) {
            return "candidate tire pressure front-left";
        }
        if ("22248F".equals(clean)) {
            return "candidate tire pressure front-right";
        }
        if ("222490".equals(clean)) {
            return "candidate tire pressure rear-right";
        }
        if ("222491".equals(clean)) {
            return "candidate tire pressure rear-left";
        }
        if ("22C901".equals(clean)) {
            return "candidate tire pressures";
        }
        if ("22C902".equals(clean)) {
            return "candidate tire temperatures";
        }
        if ("224051".equals(clean)
                || "224052".equals(clean)
                || "224053".equals(clean)
                || "224054".equals(clean)) {
            return "candidate tire receiver slot " + clean.substring(clean.length() - 1);
        }
        if ("0902".equals(clean)) {
            return "vin";
        }
        return "";
    }

    static void appendProbeLine(StringBuilder raw, String label, String value) {
        raw.append(label).append(": ").append(value == null ? "" : value).append('\n');
    }

    static String tail(String value, int maxLength) {
        if (value == null || value.length() <= maxLength) {
            return value;
        }
        return value.substring(value.length() - maxLength);
    }

    /** Exponential backoff between OBD reconnect attempts, capped at 30 s. */
    static long reconnectBackoffMs(int attempt) {
        if (attempt < 1) {
            return 0L;
        }
        long base = 2000L * (1L << Math.min(attempt - 1, 4));
        return Math.min(30000L, base);
    }

    /**
     * Backoff before retrying the initial connect, before any link was ever established. Quicker
     * and gentler than {@link #reconnectBackoffMs}: a first-attempt RFCOMM glitch retries almost
     * immediately, and a genuinely asleep car exhausts its tries fast instead of stalling the user
     * through ~90 s of exponential backoff.
     */
    static long initialConnectBackoffMs(int attempt) {
        if (attempt < 1) {
            return 0L;
        }
        return Math.min(3000L, 500L * attempt);
    }

    static String finishStatusFor(String state) {
        if ("error".equals(state) || "blocked".equals(state)) {
            return ObdLocalStore.STATUS_ERROR;
        }
        // "complete" means the session actually reached the adapter. A teardown while
        // still "connecting"/"initializing" — e.g. the user retried before the link
        // came up — never connected, so it falls through to "disconnected" below.
        if ("connected".equals(state)
                || "scanning".equals(state)
                || "scan-complete".equals(state)) {
            return ObdLocalStore.STATUS_COMPLETE;
        }
        return ObdLocalStore.STATUS_DISCONNECTED;
    }

    static String friendlyConnectionMessage(Exception ex) {
        String message = safeMessage(ex).toLowerCase(Locale.US);
        if (message.contains("socket might closed")
                || message.contains("timeout")
                || message.contains("read failed")) {
            return "Adapter serial channel did not open. Make sure the car is awake, close other OBD apps, then retry.";
        }
        if (message.contains("permission")) {
            return "Bluetooth permission is missing. Grant permissions, then retry.";
        }
        return "OBD connection failed: " + safeMessage(ex);
    }

    static String safeMessage(Exception ex) {
        String message = ex.getMessage();
        if (message == null || message.trim().isEmpty()) {
            return ex.getClass().getSimpleName();
        }
        return message;
    }

    static double round1(double value) {
        return Math.round(value * 10.0) / 10.0;
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/ObdElmDecode.kt
package com.volttracker.obdpoc

import com.volttracker.obdpoc.data.ObdLocalStore
import java.util.Locale

/**
 * Stateless decoding, classification and formatting helpers for the OBD/ELM layer.
 */
object ObdElmDecode {
    @JvmStatic
    fun hasElmPrompt(response: String?): Boolean = response != null && response.indexOf('>') >= 0

    /**
     * True when an ELM reply carries no usable OBD payload for the PID that was asked — the adapter
     * answered "NO DATA" or returned only a prompt / whitespace. The live-poll negative-PID cache
     * uses this to stop re-issuing PIDs the car never answers, and the engine uses it to notice a
     * fully asleep bus.
     *
     * Deliberately narrow: transient "CAN ERROR" / "STOPPED" / "BUFFER FULL" frames all carry
     * hex-ish characters and are NOT treated as no-data, so a momentary bus hiccup can never disable
     * a PID that is actually supported. ("NO DATA" itself contains hex letters A/D, so the literal
     * check must come before the no-hex fallback.)
     */
    @JvmStatic
    fun isNoDataResponse(response: String?): Boolean {
        val cleaned = (response ?: "").uppercase(Locale.US)
        if (cleaned.contains("NO DATA")) {
            return true
        }
        return cleaned.none { it in '0'..'9' || it in 'A'..'F' }
    }

    @JvmStatic
    fun appendRaw(
        raw: String,
        command: String,
        response: String?,
    ): String = raw + command + ": " + summarizeForStorage(command, response) + "\n"

    @JvmStatic
    fun summarizeForStorage(
        command: String?,
        response: String?,
    ): String {
        val summary = ObdProtocol.summarize(response)
        if (!isVinCommand(command)) {
            return summary
        }
        if (summary.isEmpty()) {
            return ""
        }
        return "[VIN redacted; responseLength=${summary.length}]"
    }

    private fun isVinCommand(command: String?): Boolean =
        command != null && command.trim().uppercase(Locale.US) == "0902"

    @JvmStatic
    fun pidForCommand(command: String?): String {
        if (command == null) {
            return ""
        }
        val clean = command.trim().uppercase(Locale.US)
        if (clean.startsWith("01") && clean.length >= 4) {
            return clean.substring(2, 4)
        }
        if (clean.startsWith("22") && clean.length >= 6) {
            return clean.substring(2)
        }
        if (clean.startsWith("09") && clean.length >= 4) {
            return clean.substring(2, 4)
        }
        return ""
    }

    @JvmStatic
    fun nameForCommand(command: String?): String {
        if (command == null) {
            return ""
        }
        val clean = command.trim().uppercase(Locale.US)
        return when (clean) {
            "ATRV" -> "adapter voltage"
            "010D" -> "vehicle speed"
            "010C" -> "engine rpm"
            "0105" -> "coolant temperature"
            "010F" -> "intake air temperature"
            "0104" -> "engine load"
            "0111" -> "throttle position"
            "0149" -> "accelerator pedal position"
            "0142" -> "control module voltage"
            "011F" -> "engine run time"
            "01A6" -> "odometer"
            "012F" -> "fuel level"
            "015C" -> "engine oil temperature"
            "22119F", "22119F01" -> "engine oil life"
            "221154" -> "engine oil temperature"
            "22203F" -> "engine torque"
            "015B" -> "state of charge"
            "222429" -> "hv pack voltage"
            "222414" -> "hv pack current"
            "222883" -> "motor A current"
            "222884" -> "motor B current"
            "222885" -> "motor A voltage"
            "222886" -> "motor B voltage"
            "222487" -> "ev distance this cycle"
            "222889" -> "prndl state"
            "221940", "22194001" -> "transmission temperature"
            "22434F" -> "hv battery temperature"
            "224368" -> "charger ac voltage"
            "224369" -> "charger ac current"
            "22436B" -> "charger hv voltage"
            "22436C" -> "charger hv current"
            "224373" -> "charging mode"
            "22437D" -> "last charge energy"
            "2243A5" -> "hv battery charge count"
            "2243AF" -> "hv battery raw soc"
            "2241A3" -> "hv battery capacity"
            "2245F9" -> "hv battery capacity fallback"
            "224329" -> "minimum cell voltage"
            "22432A" -> "minimum cell number"
            "22432B" -> "maximum cell voltage"
            "22432C" -> "maximum cell number"
            "22435F" -> "SOC variation"
            "2240E9" -> "pack resistance"
            "22433B" -> "minimum pack voltage"
            "22433C" -> "maximum pack voltage"
            "224349" -> "hv battery max temperature"
            "22434A" -> "hv battery min temperature"
            "22434B" -> "hv battery max-temp module"
            "22434C" -> "hv battery min-temp module"
            "221C43" -> "power electronics coolant loop temperature"
            "2241A4" -> "battery coolant temperature"
            "22433F" -> "minimum SOC limit"
            "2241B0" -> "APM output power"
            "22437E" -> "APM output current"
            "2243A6" -> "HV isolation resistance (kOhm)"
            "2241EC" -> "HV isolation resistance (ohm)"
            "2241B1" -> "AC compressor commanded power"
            "2241B3" -> "cabin heater commanded power"
            "2241B5" -> "battery heater commanded power"
            "2282B5" -> "AC compressor speed"
            "2282B7" -> "AC compressor power"
            "224531" -> "charging level"
            "228334" -> "hv battery displayed soc"
            "2241B2" -> "battery coolant pump rpm"
            "2241B4" -> "battery coolant valve"
            "2241B6" -> "battery heater power"
            "22801E" -> "outside air temperature raw"
            "22801F" -> "outside air temperature filtered"
            "221141" -> "ignition / 12V voltage"
            "221C47" -> "14V setpoint voltage"
            "221C26" -> "inverter temperature 1"
            "221C28" -> "inverter temperature 2"
            "221C2A" -> "inverter temperature 3"
            "2228CB" -> "motor temperature"
            "22242C" -> "brake torque demand"
            "2224B0" -> "regen braking active"
            "224501" -> "brake pedal position 1"
            "224502" -> "brake pedal position 2"
            "2240D7" -> "battery section 1 temperature"
            "2240D9" -> "battery section 2 temperature"
            "2240DB" -> "battery section 3 temperature"
            "2240DD" -> "battery section 4 temperature"
            "2240DF" -> "battery section 5 temperature"
            "2240E1" -> "battery section 6 temperature"
            "2240D4" -> "HD pack current"
            "22C218" -> "average cell voltage"
            "0132" -> "EVAP vapor pressure"
            "22248E" -> "candidate tire pressure front-left"
            "22248F" -> "candidate tire pressure front-right"
            "222490" -> "candidate tire pressure rear-right"
            "222491" -> "candidate tire pressure rear-left"
            "22C901" -> "candidate tire pressures"
            "22C902" -> "candidate tire temperatures"
            "224051", "224052", "224053", "224054" -> "candidate tire receiver slot ${clean.substring(
                clean.length - 1,
            )}"
            "0902" -> "vin"
            else -> ""
        }
    }

    @JvmStatic
    fun appendProbeLine(
        raw: StringBuilder,
        label: String,
        value: String?,
    ) {
        raw
            .append(label)
            .append(": ")
            .append(value ?: "")
            .append('\n')
    }

    @JvmStatic
    fun tail(
        value: String?,
        maxLength: Int,
    ): String? {
        if (value == null || value.length <= maxLength) {
            return value
        }
        return value.substring(value.length - maxLength)
    }

    /** Exponential backoff between OBD reconnect attempts, capped at 30 s. */
    @JvmStatic
    fun reconnectBackoffMs(attempt: Int): Long {
        if (attempt < 1) {
            return 0L
        }
        val base = 2000L * (1L shl minOf(attempt - 1, 4))
        return minOf(30000L, base)
    }

    @JvmStatic
    fun initialConnectBackoffMs(attempt: Int): Long {
        if (attempt < 1) {
            return 0L
        }
        return minOf(3000L, 500L * attempt)
    }

    @JvmStatic
    fun finishStatusFor(state: String?): String {
        if (state == "error" || state == "blocked") {
            return ObdLocalStore.STATUS_ERROR
        }
        if (state == "connected" || state == "scanning" || state == "scan-complete") {
            return ObdLocalStore.STATUS_COMPLETE
        }
        return ObdLocalStore.STATUS_DISCONNECTED
    }

    @JvmStatic
    fun friendlyConnectionMessage(ex: Exception): String {
        val message = safeMessage(ex).lowercase(Locale.US)
        if (message.contains("socket might closed") ||
            message.contains("timeout") ||
            message.contains("read failed")
        ) {
            return "Adapter serial channel did not open. Make sure the car is awake, close other OBD apps, then retry."
        }
        if (message.contains("permission")) {
            return "Bluetooth permission is missing. Grant permissions, then retry."
        }
        return "OBD connection failed: ${safeMessage(ex)}"
    }

    @JvmStatic
    fun safeMessage(ex: Exception): String {
        val message = ex.message
        if (message == null || message.trim().isEmpty()) {
            return ex.javaClass.simpleName
        }
        return message
    }

    @JvmStatic
    fun round1(value: Double): Double = Math.round(value * 10.0) / 10.0
}
